### PR TITLE
[msan] Add 'MappingDesc::ALLOCATOR' type and check it is available

### PR DESCRIPTION
**Pull request by thurstond**
MSan divides the virtual address space into APP, INVALID, SHADOW and ORIGIN memory. The allocator usually just steals a bit of the APP address space: typically the bottom portion of the PIE binaries section, which works because the Linux kernel maps from the top of the PIE binaries section. However, if ASLR is very aggressive, the binary may end up mapped in the same location where the allocator wants to live; this results in a segfault.

This patch adds in a MappingDesc::ALLOCATOR type and enforces that the memory range for the allocator is not occupied by anything else.

Since the allocator range information is not readily available in msan.h, we duplicate the information from msan_allocator.cpp.

Note: aggressive ASLR can also lead to a different type of failure, where the PIE binaries/libraries are mapped entirely outside of the APP/ALLOCATOR sections; that will be addressed in a separate patch (https://github.com/llvm/llvm-project/pull/85142).

### CODE REVIEW

File: compiler-rt/lib/msan/msan.h [hunk 1]
```diff
@@ -32,13 +32,13 @@
 struct MappingDesc {
   uptr start;
   uptr end;
-  enum Type {
-    INVALID, APP, SHADOW, ORIGIN
-  } type;
+  enum Type { INVALID, ALLOCATOR, APP, SHADOW, ORIGIN } type;
```
Comment (vitalybuka) [inline compiler-rt/lib/msan/msan.h:35]:
```suggestion
  enum Type { 
    INVALID = 1,
    ALLOCATOR  =2,
    APP = 4,
    SHADOW = 8,
    ORIGIN = 16,
   } type;
```

File: compiler-rt/lib/msan/msan.h [hunk 2]
```diff
@@ -236,7 +241,9 @@ inline bool addr_is_type(uptr addr, MappingDesc::Type mapping_type) {
   return false;
 }
 
-#define MEM_IS_APP(mem) addr_is_type((uptr)(mem), MappingDesc::APP)
+#define MEM_IS_APP(mem)                           \
```
Comment (vitalybuka) [inline compiler-rt/lib/msan/msan.h:244]:
```suggestion
#define MEM_IS_APP(mem) addr_is_type((uptr)(mem), MappingDesc::APP | MappingDesc::ALLOCATOR))
```

Comment (vitalybuka) [reply compiler-rt/lib/msan/msan.h:244]: with appropriate update of addr_is_type

**Review by vitalybuka — APPROVED** [review]

File: compiler-rt/lib/msan/msan.h (hunk 2 above)
Comment (thurstond) [reply compiler-rt/lib/msan/msan.h]: Thanks, done!

**Review by vitalybuka — APPROVED** [review]

**Pull request merged**